File: Ruby/Basico/gestor_archivos/gestor.rb
require 'fileutils'

def listar(dir)
  Dir.entries(dir).reject { |f| f.start_with?('.') }.each { |f| puts f }
end

def crear_archivo(ruta, contenido = '')
  File.write(ruta, contenido)
  puts "Archivo creado: #{ruta}"
end

def copiar(origen, destino)
  FileUtils.cp(origen, destino)
  puts "Copiado: #{origen} → #{destino}"
end

def mover(origen, destino)
  FileUtils.mv(origen, destino)
  puts "Movido: #{origen} → #{destino}"
end

def eliminar(ruta)
  File.delete(ruta)
  puts "Eliminado: #{ruta}"
rescue Errno::ENOENT
  puts "Archivo no encontrado: #{ruta}"
end

def buscar(dir, patron)
  Dir.glob("#{dir}/**/#{patron}").each { |f| puts f }
end

puts '=== Gestor de Archivos ==='
loop do
  puts "\n1. Listar  2. Crear  3. Copiar  4. Mover  5. Eliminar  6. Buscar  7. Salir"
  print 'Opción: '
  case gets.chomp
  when '1'
    print 'Directorio: '; listar(gets.chomp)
  when '2'
    print 'Ruta: '; ruta = gets.chomp
    print 'Contenido: '; crear_archivo(ruta, gets.chomp)
  when '3'
    print 'Origen: '; o = gets.chomp; print 'Destino: '; copiar(o, gets.chomp)
  when '4'
    print 'Origen: '; o = gets.chomp; print 'Destino: '; mover(o, gets.chomp)
  when '5'
    print 'Ruta: '; eliminar(gets.chomp)
  when '6'
    print 'Directorio: '; d = gets.chomp; print 'Patrón (ej: *.rb): '; buscar(d, gets.chomp)
  when '7'
    break
  end
end

File: PHP/Basico/gestor_archivos/gestor.php
<?php

function listar(string $dir): void
{
    foreach (scandir($dir) as $item) {
        if (!str_starts_with($item, '.')) echo "$item\n";
    }
}

function crear(string $ruta, string $contenido = ''): void
{
    file_put_contents($ruta, $contenido);
    echo "Creado: $ruta\n";
}

function copiar(string $origen, string $destino): void
{
    copy($origen, $destino);
    echo "Copiado: $origen → $destino\n";
}

function mover(string $origen, string $destino): void
{
    rename($origen, $destino);
    echo "Movido: $origen → $destino\n";
}

function eliminar(string $ruta): void
{
    if (file_exists($ruta)) {
        unlink($ruta);
        echo "Eliminado: $ruta\n";
    } else {
        echo "No encontrado: $ruta\n";
    }
}

function buscar(string $dir, string $patron): void
{
    foreach (glob("$dir/**/$patron", GLOB_BRACE) as $archivo) {
        echo "$archivo\n";
    }
}

echo "=== Gestor de Archivos ===\n";

while (true) {
    echo "\n1. Listar  2. Crear  3. Copiar  4. Mover  5. Eliminar  6. Buscar  7. Salir\n";
    echo "Opción: ";
    $op = trim(fgets(STDIN));

    match ($op) {
        '1' => (function () { echo 'Directorio: '; listar(trim(fgets(STDIN))); })(),
        '2' => (function () { echo 'Ruta: '; $r = trim(fgets(STDIN)); echo 'Contenido: '; crear($r, trim(fgets(STDIN))); })(),
        '3' => (function () { echo 'Origen: '; $o = trim(fgets(STDIN)); echo 'Destino: '; copiar($o, trim(fgets(STDIN))); })(),
        '4' => (function () { echo 'Origen: '; $o = trim(fgets(STDIN)); echo 'Destino: '; mover($o, trim(fgets(STDIN))); })(),
        '5' => (function () { echo 'Ruta: '; eliminar(trim(fgets(STDIN))); })(),
        '6' => (function () { echo 'Directorio: '; $d = trim(fgets(STDIN)); echo 'Patrón: '; buscar($d, trim(fgets(STDIN))); })(),
        '7' => exit(0),
        default => print("Opción inválida\n"),
    };
}
